Duxiu.js

```diff
@@ -0,0 +1,423 @@
+		pattern = /页[\s]*数\D*([\s\S]*?)<\/dd>/;
+		if (pattern.test(page)) {
+			var numPages = pattern.exec(page)[1];
+			newItem.numPages = Zotero.Utilities.trim(numPages);
+		}
+		
+		// 丛书 book series.
+		pattern = /<dd>[\s\S]*丛书名[\s\S]*?>([\s\S]*?)<\/dd>/;
+		if (pattern.test(page)) {
+			var series = trimTags(pattern.exec(page)[1]);
+			newItem.series = Zotero.Utilities.trim(series);
+		}
+		// 原书定价 retail price.
+		pattern = /原书定价\D*([\s\S]*?)<\/dd>/;
+		if (pattern.test(page)) {
+			var price = pattern.exec(page)[1];
+			newItem.price = Zotero.Utilities.trim(price);
+			newItem.extra += "原书定价: " + newItem.price + "\n";
+		}
+		
+		// 开本 edition format.
+		pattern = /<dd>[\s\S]*开本[\s\S]*?>([\s\S]*?)<\/dd>/;
+		if (pattern.test(page)) {
+			var format = trimTags(pattern.exec(page)[1]);
+			newItem.format = Zotero.Utilities.trim(format);
+			newItem.extra += "开本: " + newItem.format + "\n";
+		}
+		// 主题词 subject terms.
+		pattern = /<dd>[\s\S]*主题词[\s\S]*?>([\s\S]*?)<\/dd>/;
+		if (pattern.test(page)) {
+			var subjectTerms = trimTags(pattern.exec(page)[1]);
+			newItem.subjectTerms = Zotero.Utilities.trim(subjectTerms);
+		}
+	
+		// 中图法分类号 CLC classification number.
+		pattern = /<dd>[\s\S]*中图法分类号[\s\S]*?>([\s\S]*?)<\/dd>/;
+		if (pattern.test(page)) {
+			var callNumber = trimTags(pattern.exec(page)[1]);
+			newItem.callNumber = Zotero.Utilities.trim(callNumber);
+		}		
+		
+		// 参考文献格式 reference format.
+		pattern = /<dd>[\s\S]*参考文献格式[\s\S]*?>([\s\S]*?)<\/dd>/;
+		if (pattern.test(page)) {
+			var refFormat = trimTags(pattern.exec(page)[1]);
+			newItem.refFormat = Zotero.Utilities.trim(refFormat);
+			
+			newItem.extra = "参考格式: " + newItem.refFormat + "\n" + newItem.extra;
+		}
+		
+		// 内容提要 abstract.
+		pattern = /<dd>[\s\S]*内容提要[\s\S]*?>([\s\S]*?)<\/dd>/;
+		if (pattern.test(page)) {
+			var abstractNote = trimTags(pattern.exec(page)[1]);
+			newItem.abstractNote = Zotero.Utilities.trim(abstractNote).replace(/&mdash;/g, "-") + "\n\n";
+		}
+		
+		// use subject terms to populate abstract
+		if (newItem.subjectTerms) {
+			newItem.abstractNote = newItem.abstractNote + "主题词: " + newItem.subjectTerms;
+		}
+			
+		// start the abstract with the foreign language title if available.
+		if (newItem.foreignTitle) {
+			newItem.abstractNote = "外文题名: " + newItem.foreignTitle + "\n\n" + newItem.abstractNote;
+		}
+
+		// SSID
+		pattern = /<input name = "ssid" id = "forumssid" {2}value = "([\s\S]*?)"/;
+		if (pattern.test(page)) {
+			var SSID = trimTags(pattern.exec(page)[1]);
+			newItem.SSID = Zotero.Utilities.trim(SSID);
+			newItem.extra = newItem.extra + "SSID: " + newItem.SSID;
+		}
+
+		newItem.complete();
+	});
+}
+
+// the list from which to pick the best role for a given creator. Do not add variants of strings that end with 著,译，编
+var rolelist = ["总主编", "总编辑", "总编", "编著", "编译", "编", "整理", "执笔", "译", "著", "撰", "纂", "集解", "辑", "编辑", "集注","绘"];
+
+function trimTags(text) {
+	return text.replace(/(<.*?>)|\t|\r|(隐藏更多)|&nbsp;|/g, "");
```

Comment:
Look at `ZU.cleanTags()` and maybe also `ZU.trimInternal()` which should to a similar thing.

Reply:
Thanks so much @zuphilip ! I can make the corrections later today or tomorrow (such as double calling the website and unnecessary lines). But I do have a follow-up question about using RegEx:

I realized using all regex is not as stable, but the problem I had is that the Duxiu entry html structure is messy. The info section is like (if the Chinese characters are translated into English):
```
<dl>
<dt>TITLE</dt>
<dd>
PUBLICATION DATE : 2011.10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
P&nbsp;A&nbsp;G&nbsp;E : 549</dd>
<dd>SERIES</dd>
<dd>ISBN: 12345678910</dd>
...
</dl>
```
I know I can use xPath to pick out `<dt>` (the title), but can I do it with the `<dd>` (all the other entries)? 

Or should I mix xPath and regex? Like pick out `<dl>` first and then parse all the `<dd>` using shorter regex?

This point was actually raised earlier on the forum by someone else: https://forums.zotero.org/discussion/76178/can-the-zotero-connector-catch-the-information-of-duxiu
And I quote from adamsmith:
> but as a data structure, it's a complete mess: all of this is in unnamed HTML nodes, so the only way to get to it is by parsing plain text using regular expressions, which is super fragile. It's doable, but it's not something I prioritize given how easily it breaks.

Which is why I started with all regex. But I would be happy to switch to xPath if possible.

And to answer your earlier question:
1. The database can be accessed fully with subscription and partially (few functionalities) with "trial option". To use trial option, it asks you to click on a button (the button below on the right) to store the session. This can be done in a sys browser, but not always in Scaffold browser. 
![image](https://user-images.githubusercontent.com/7269717/69409316-f0425580-0d08-11ea-8597-43bb10f5f08c.png)

 2. In practice, it has been working alright for the past week in daily uses, and I have added support for journal articles. I'll wait for your comment on the question above before making another commit.

Thanks again!

Reply:
Thank you very much! I click on one of the links below the blue box and can enter the database as well (for me the link was on the left). That is useful, as maintenance of this translator is then also easier.

Yes, a mix with xpath and regular expression might be the way to go here. On a general level I think it should be easy to get all the main information parts with xpaths and decide what they are about. The more fine-grained information could then be parsed with some regex, i.e. something like
```js
var informations= doc.querySelectorAll('dl>dd');
for (let info of informations) {
  let content = info.textContent;
  if (!content.includes(':')) continue;
  let field = content.split(':')[0].trim();
  let value = content.split(':').shift().join(':').trim();
  // information about publisher, place, year
  // e.g. 出版发行 : 上海：中华书局 , 1946.09
  if (field === "出版发行") {
    let m = value.match(...);
...
```
(Note, this code is not tested)

Let me know if you have more questions or when this is ready for another review.